src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -204,1024 +204,1040 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
+                        requestParams.RequestContext.Logger.Info(
+                            "Access token is not expired. Returning the found cache entry. " +
+                            GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
+                        return msalAccessTokenCacheItem;
+                    }
+
+                    if (requestParams.IsExtendedLifeTimeEnabled && msalAccessTokenCacheItem.ExtendedExpiresOn >
+                        DateTime.UtcNow + TimeSpan.FromMinutes(DefaultExpirationBufferInMinutes))
+                    {
+                        requestParams.RequestContext.Logger.Info(
+                            "Access token is expired.  IsExtendedLifeTimeEnabled=TRUE and ExtendedExpiresOn is not exceeded.  Returning the found cache entry. " +
+                            GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
+
+                        msalAccessTokenCacheItem.IsExtendedLifeTimeToken = true;
+                        return msalAccessTokenCacheItem;
+                    }
+
                     requestParams.RequestContext.Logger.Info(
-                        "Access token is expired.  IsExtendedLifeTimeEnabled=TRUE and ExtendedExpiresOn is not exceeded.  Returning the found cache entry. " +
+                        "Access token has expired or about to expire. " +
                         GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
-
-                    msalAccessTokenCacheItem.IsExtendedLifeTimeToken = true;
-                    return msalAccessTokenCacheItem;
                 }
 
-                requestParams.RequestContext.Logger.Info(
-                    "Access token has expired or about to expire. " +
-                    GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
+                return null;
             }
-
-            return null;
         }
-    }
 
-    private string GetAccessTokenExpireLogMessageContent(MsalAccessTokenCacheItem msalAccessTokenCacheItem)
-    {
-        return string.Format(
-            CultureInfo.InvariantCulture,
-            "[Current time ({0}) - Expiration Time ({1}) - Extended Expiration Time ({2})]",
-            DateTime.UtcNow,
-            msalAccessTokenCacheItem.ExpiresOn,
-            msalAccessTokenCacheItem.ExtendedExpiresOn);
-    }
+        private ICollection<MsalAccessTokenCacheItem> FilterToAuthoritySpecifiedByAuthenticationRequest(
+            AuthenticationRequestParameters requestParams, ICollection<MsalAccessTokenCacheItem> tokenCacheItems)
+        {
+            requestParams.RequestContext.Logger.Info("Filtering by authority specified in the authentication request parameters...");
+            return tokenCacheItems
+                    .Where(item => item.Authority.Equals(requestParams.Authority.CanonicalAuthority,
+                                                         StringComparison.OrdinalIgnoreCase))
+                    .ToList();
```

Comment:
Consider making the filter and only update the tokenCacheItems in case there is at least one result?

Reply:
clarifying. pls. assign the result to a different variable. if there are no items found return the tokenCacheItems unfiltered. If there is more than one item return the filtered items. This will ensure the fix is as defensive as possible.

---
In reply to: [239592260](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/708#discussion_r239592260) [](ancestors = 239592260)

Reply:
thanks for the clarification. much appreciated

---
In reply to: [239635201](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/708#discussion_r239635201) [](ancestors = 239635201,239592260)